wallet-api/test/Generators.hs

```diff
@@ -0,0 +1,108 @@
+{-# OPTIONS -fplugin=Language.Plutus.CoreToPLC.Plugin -fplugin-opt Language.Plutus.CoreToPLC.Plugin:dont-typecheck #-}
+module Generators(
+    -- * Mockchain
+    Mockchain(..),
+    genMockchain,
+    -- * Transactions
+    genInitialTransaction,
+    genValidTransaction,
+    -- * Useful scripts
+    unitData,
+    unitRedeemer,
+    emptyValidator
+    ) where
+
+import           Crypto.Hash           (Digest, SHA256, hash)
+import           Data.Map              (Map)
+import qualified Data.Map              as Map
+import           Data.Maybe (fromMaybe)
+import           Hedgehog
+import qualified Hedgehog.Gen          as Gen
+import qualified Hedgehog.Range        as Range
+
+import           Language.Plutus.TH    (plutusT)
+
+import           Wallet.Emulator.Types
+import           Wallet.UTXO
+
+-- | Blockchain for testing the emulator implementation and traces.
+--   
+--   To avoid having to rely on functions from the implementation of
+--   wallet-api (in particular, `Wallet.UTXO.unspentOutputs`) we note the
+--   unspent outputs of the chain when it is first created.
+data Mockchain = Mockchain {
+    mockchainBlockchain :: Blockchain,
+    mockchainUtxo :: Map TxOutRef TxOut
+    } deriving Show
+
+-- | Generate a mockchain
+--
+--   TODO: Generate more than 1 txn
+genMockchain :: MonadGen m
+    => m Mockchain
+genMockchain = do
+    (txn, ot) <- genInitialTransaction
+    let txId = hashTx txn
+    pure $ Mockchain {
+        mockchainBlockchain = [[txn]],
+        mockchainUtxo = Map.singleton (TxOutRef txId 0) ot
+        }
+
+-- | A transaction with no inputs that forges some value (to be used at the 
+--   beginning of a blockchain)
+genInitialTransaction :: MonadGen m
+    => m (Tx, TxOut)
+genInitialTransaction = do
+    vl <- Value <$> Gen.integral (Range.linear 1 1000000)
+    let o = simpleOutput vl
+    pure (Tx {
+        txInputs = [],
+        txOutputs = [o],
+        txForge = vl,
+        txFee = 0
+        }, o)
+
+-- | Transaction output locked by the empty validator and unit data scripts.
+simpleOutput :: Value -> TxOut
+simpleOutput vl = TxOut (hashValidator emptyValidator) vl unitData
+
+-- | Generate a valid transaction, using the unspent outputs provided.
+--   Fails if the there are no unspent outputs, or if the total value
+--   of the unspent outputs is smaller than the minimum fee (1)
+genValidTransaction :: MonadGen m
+    => Blockchain
+    -> Map TxOutRef TxOut
+    -> m Tx
+genValidTransaction bc ops = do
+    -- Take a random number of UTXO from the input
+    nUtxo <- if Map.null ops
+             then Gen.discard
+             else Gen.int (Range.linear 1 (Map.size ops))
+    let inputs = fmap (\ref -> TxIn ref emptyValidator unitRedeemer)
+            $ take nUtxo 
+            $ fmap fst
+            $ Map.toList ops
+        totalVal = sum (map (fromMaybe 0 . value bc . txInRef) inputs)
+        minFee = 1
+    fee <- Gen.integral (Range.linear minFee totalVal)
+    if fee < totalVal
```

Comment:
how can it not be given the bounds you put in the generator?

Reply:
The range is *[inclusive, inclusive]* so it could be that `fee == totalVal`

Reply:
Hm, I wonder if the range generators fail if they're given an empty range? If so, you could simplify some of the things you do above with the map size and here by doing `Range.linear minFee (totalVal-1)`. Otherwise this seems fine.

Reply:
> I wonder if the range generators fail if they're given an empty range?

According to the hedgehog [source ](https://hackage.haskell.org/package/hedgehog-0.6.1/docs/src/Hedgehog.Internal.Seed.html#nextInteger) it calls [System.Random.randomR](http://hackage.haskell.org/package/random-1.1/docs/System-Random.html) and its docs say 

>  It is unspecified what happens if lo>hi.

So I think the manual check has to stay